tools/codegen/dest/register_dispatch_key.py

```diff
@@ -88,12 +113,38 @@ def gen_structured(self, g: NativeFunctionsGroup) -> List[str]:
             self.target,
             self.selector,
             self.rocm,
+            self.cpp_namespace,
             g
         )
         return list(mapMaybe(structured_gen.gen_one, g.functions()))
 
-    @method_with_native_function
```

Comment:
So, I'm going to have to add this back before I merge the PR. One annoying aspect of my change is that I added an optional `is_generated_wrapper` parameter to `gen_unstructured`. In order to re-use `gen_unstructured` for the code-generated inplace/out wrappers, `gen_unstructured` needs that extra bit of information: it's not available just from the `ExternalBackendFunction`, because it requires information relevant to the entire function group. i.e. we only generate a wrapper if (a) the inplace/out kernel is not implemented, and (b) the functional kernel is implemented.

Another option would have been to try to factor out the logic for registrations/definitions/declarations into their own functions that are then called by both `gen_unstructured` and the `gen_out_inplace_wrappers` code, but I tentatively found the approach I took here to be less confusing.

After some digging I couldn't find a way to make a flavor of `method_with_native_function` that properly deals with defaultable arguments, so my plan was to add a `with native_function_manager(f): ...` at the beginning of this function. That would indent the entire function and make it impossible to see the delta in this PR though, so my plan is to wait until it's been reviewed to add that in. Or, take it out completely depending on what the feedback is :)

Reply:
I didn't understand all of this (still reading), but indentation is not a problem, you can tell people to use GH ignore whitespace feature (or review on phab)

Reply:
Probably just need to rework `@method_with_native_function` entirely. Tis really annoying to do decorators and strongly typed Python at the same time.